runtime/src/axum/mod.rs

```diff
@@ -0,0 +1,529 @@
+}
+
+impl AxumWasm {
+    pub fn new() -> Self {
+        // Allow about 2^15 = 32k logs of backpressure
+        // We know the wasm currently handles about 16k requests per second (req / sec) so 16k seems to be a safe number
+        // As we make performance gains elsewhere this might eventually become the new bottleneck to increase :D
+        //
+        // Testing has shown that a number half the req / sec yields poor performance. A number the same as the req / sec
+        // seems acceptable so going with double the number for some headroom
+        let (tx, rx) = mpsc::channel(1 << 15);
+
+        Self {
+            router: Mutex::new(None),
+            logs_rx: Mutex::new(Some(rx)),
+            logs_tx: Mutex::new(tx),
+            kill_tx: Mutex::new(None),
+        }
+    }
+}
+
+impl Default for AxumWasm {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[async_trait]
+impl Runtime for AxumWasm {
+    async fn load(
+        &self,
+        request: tonic::Request<LoadRequest>,
+    ) -> Result<tonic::Response<LoadResponse>, Status> {
+        let wasm_path = request.into_inner().path;
+        trace!(wasm_path, "loading");
+
+        let router = RouterBuilder::new()
+            .map_err(|err| Status::from_error(err.into()))?
+            .src(wasm_path)
+            .build()
+            .map_err(|err| Status::from_error(err.into()))?;
+
+        *self.router.lock().unwrap() = Some(router);
+
+        let message = LoadResponse { success: true };
+
+        Ok(tonic::Response::new(message))
+    }
+
+    async fn start(
+        &self,
+        request: tonic::Request<StartRequest>,
+    ) -> Result<tonic::Response<StartResponse>, Status> {
+        let StartRequest {
+            deployment_id,
+            port,
+            ..
+        } = request.into_inner();
+
+        let address = SocketAddr::new(Ipv4Addr::LOCALHOST.into(), port as u16);
+
+        let logs_tx = self.logs_tx.lock().unwrap().clone();
+
+        let (kill_tx, kill_rx) = tokio::sync::oneshot::channel();
+
+        *self.kill_tx.lock().unwrap() = Some(kill_tx);
+
+        let router = self
+            .router
+            .lock()
+            .unwrap()
+            .take()
+            .context("tried to start a service that was not loaded")
+            .map_err(|err| Status::internal(err.to_string()))?;
+
+        tokio::spawn(run_until_stopped(
+            router,
+            deployment_id,
+            address,
+            logs_tx,
+            kill_rx,
+        ));
+
+        let message = StartResponse { success: true };
+
+        Ok(tonic::Response::new(message))
+    }
+
+    type SubscribeLogsStream = ReceiverStream<Result<runtime::LogItem, Status>>;
+
+    async fn subscribe_logs(
+        &self,
+        _request: tonic::Request<SubscribeLogsRequest>,
+    ) -> Result<tonic::Response<Self::SubscribeLogsStream>, Status> {
+        let logs_rx = self.logs_rx.lock().unwrap().deref_mut().take();
```

Comment:
Missing comment for why `unwrap` is safe. Do these even need to be behind a mutex?